lib/src/json/sink/sink.dart

```diff
@@ -0,0 +1,108 @@
+///
+/// In general, a [JsonSink] is not required or expected to
+/// work correctly if calls are performed out of order.
+/// Only call sequences corresponding to a correct JSON structure
+/// are guaranteed to give a meaningful result.
+abstract class JsonSink {
+  /// Called for a number value.
+  void addNumber(num value);
+
+  /// Called for a null value.
+  void addNull();
+
+  /// Called for a string value.
+  void addString(String value);
+
+  /// Called for a boolean value.
+  void addBool(bool value);
+
+  /// Called at the beginning of an array value.
+  ///
+  /// Each value added until a corresponding [endArray]
+  /// is considered an element of this array, unless it's part of a nested
+  /// array or object.
+  void startArray();
+
+  /// Ends the current array.
+  ///
+  /// The array value is now complete, and should
+  /// be treated as a value of a surrounding array or obejct.
+  void endArray();
+
+  /// Called at the beginning of an object value.
+  ///
+  /// Each value added until a corresponding [endObject]
+  /// is considered an entry value of this object, unless it's part of a nested
+  /// array or object.
+  /// Each such added value must be preceeded by exactly one call to [addKey]
+  /// which provides the corresponding key.
+  void startObject();
+
+  /// Sets the key for the next value of an object.
+  ///
+  /// Should preceed any value or array/object start inside an object.
+  void addKey(String key);
+
+  /// Ends the current object.
+  ///
+  /// The object value is now complete, and should
+  /// be treated as a value of a surrounding array or obejct.
+  void endObject();
+}
+
+/// Creates a [JsonSink] which builds a JSON string.
+///
+/// The string is written to [sink].
+///
+/// If [indent] is supplied, the resulting string will be "pretty printed"
+/// with array and object entries on lines of their own
+/// and indented by multiples of the [indent] string.
+/// If the [indent] string is not valid JSON whitespace,
+/// the result written to [sink] will not be valid JSON source.
+///
+/// If [asciiOnly] is set to `true`, string values will have all non-ASCII
+/// characters escaped. If not, only control characters, quotes and backslashes
+/// are escaped.
+///
+/// The returned sink is not reusable. After it has written a single JSON structure,
+/// it should not be used again.
+JsonSink jsonStringWriter(StringSink sink, {String indent, bool asciiOnly = false}) {
+  if (indent == null) return JsonStringWriter(sink, asciiOnly: asciiOnly);
+  return JsonPrettyStringWriter(sink, indent, asciiOnly: asciiOnly);
+}
+
+/// Creates a [JsonSink] which builds a Dart JSON object structure.
+///
+/// After adding values corresponding to a JSON structure to the sink,
+/// the [result] callback is called with the resulting object structure.
+///
+/// When [result] is called, the returned sink is reset and can be reused.
+JsonSink jsonObjectWriter(void Function(dynamic) result) => JsonObjectWriter(result);
```

Comment:
The API here isn't that bad. If the developer writes something like:
```dart
dynamic result;
final sink = jsonObjectWriter((r) => result = r);

// Write json to sink
writeSomeJsonToSink(sink);

// Do something with result
assert(result != null);
print(result);
```

But we risk that someone might do:
```dart
dynamic result;
final sink = jsonObjectWriter((result) {
  // Do something with result
  assert(result != null);
  print(result);  
});

// Write json to sink
writeSomeJsonToSink(sink);
```
Which could easily cause problems with try/catch, and other problems if `writeSomeJsonToSink` doesn't actually produce a result. At-least an unfortunate developer could easily create some messy code, if they don't employ a solid pattern here.

It might be wise to give consumers an example of which pattern you envision they employ when using this method (just in the doc comments).

You could also consider:
```dart
abstract class JsonSinkAndResult<T> extends JsonSink {
  /// May throw [StateError] if no value has been produced.
  /// This can probably depend on how much validation you have.
  T get result;
}
```
(but that creates another public type, I can see how that's not as elegant)

Reply:
And I'm not a fan of the "ThingAndOtherThing" classes, and I do not want sinks to expose their final value.

I'd rather go for:
```dart
Object withJsonObjectWriter(void Function(JsonSink) action) { 
   Object result;
   bool hasResult = false;
   action(jsonObjectWriter((r) {
      result = r; 
      hasResult = true;
   }));
   if (!hasResult) throw StateError("Action did not write complete value");
   return result;
}
```
used as
```
int result = withJsonObjectWriter((sink) {
   sink.startArray();
   sink.writeInt(42);
   sink.writeInt(37);
   sink.endArray();
});
```

I guess I could parameterize it with a sink factory too, but then it gets really convoluted to use.

Reply:
However, `withJsonObjectWriter` is hard to use for sync and async constructions..

Reply:
True.
The API here is inherently synchronous. A push-based API almost has to be because it has no way to recognize that an operation takes time.

I'd write a "buffering sink" if that becomes necessary, so you can buffer the events and deliver them when the sink is ready again.